packages/rest/src/coercion/coerce-parameter.ts

```diff
@@ -0,0 +1,116 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {ParameterObject, isReferenceObject} from '@loopback/openapi-v3-types';
+import {Validator} from './validator';
+import * as debugModule from 'debug';
+import {HttpErrorMessage} from '../';
+
+const debug = debugModule('loopback:rest:coercion');
+
+/**
+ * Coerce the http raw data to a JavaScript type data of a parameter
+ * according to its OpenAPI schema specification.
+ *
+ * @param data The raw data get from http request
+ * @param schema The parameter's schema defined in OpenAPI specification
+ */
+export function coerceParameter(data: string, spec: ParameterObject) {
+  const schema = spec.schema;
+  if (!schema || isReferenceObject(schema)) {
+    debug(
+      'The parameter with schema %s is not coerced since schema' +
+        'dereferrence is not supported yet.',
+      schema,
+    );
+    return data;
+  }
+  const OAIType = getOAIPrimitiveType(schema.type, schema.format);
+  const validator = new Validator({parameterSpec: spec});
+
+  validator.validateParamBeforeCoercion(data);
+
+  switch (OAIType) {
+    case 'byte':
+      return Buffer.from(data, 'base64');
+    case 'date':
+      return new Date(data);
+    case 'float':
+    case 'double':
+      return parseFloat(data);
+    case 'number':
+      const coercedData = data ? Number(data) : undefined;
+      if (coercedData === undefined) return;
+      if (isNaN(coercedData))
+        throw HttpErrorMessage.invalidData(data, spec.name);
```

Comment:
I find `HttpErrorMessage` misleading, the namespace is not for error messages anymore. How about `CommonHttpErrors` or `RestHttpErrors`?

Reply:
Maybe `RestErrors` would be best?

Reply:
Hmm..I prefer the name `RestHttpErrors`